check for cs op and odlm to be scaled after failure

bluzarraga opened the pull request:
Always check to see if the common service operator and odlm pods have been scaled back up after a failed isolate attempt. 
Scenario:
On first attempt, the isolate script successfully scales down the cs operator and odlm deployments to 0 and uninstalls cert manager. However, the script fails before it can rescale cs operator and odlm deployments so cert manager is not deployed again either. The subsequent attempt at running the isolate script will then fail immediately because no cert manager is deployed. This failure is also very obtuse as the error handling of this function is not strong enough.

What this pr does:
1. makes the early failure due to a lack of cert manager more clear
2. ensures cs operator and odlm are scaled to 1 before proceeding with the script. Wait for CS CR after scaling

Testing: 
- deploy common services
- run isolate script, make sure to exit sometime after uninstall singletons
- verify cert manager is not installed, cs op and odlm are scaled to 0
- run this version of the isolate script
- verify script completes

bluzarraga commented:
I have run the test I outlined above three times in a row successfully so I think this pr is ready for review

Daniel-Fan commented:
@bluzarraga 
hmmm I doubt that we do actually need `echo "none"`?
My assumption is that we will always have Kind `CSV` existing in the cluster, there is no non-zero status returned from first command
```
➜  csv_count=$(oc get csv -A | grep "cert-manager"| wc -l | tr -d " ")     
➜  echo $?
0
➜  echo $csv_count
2
➜  csv_count=$(oc get csv -A | grep "cert-manager-xxxx"| wc -l | tr -d " ")
➜  echo $?                                                                 
0
➜  echo $csv_count                                                         
0
```

bluzarraga commented:
Steps for test:
- ran isolate.sh, exited prematurely before cs and odlm were rescaled
- verify no cert manager and odlm and cs scaled to 0
- switch to scripts branch to run isolate.sh and get the quick error
- switch back to this branch to run fixed isolate.sh
- verified script completed and cert manager deployed
Logs from latest full problem recreation and test:
```
[root@api.critter.cp.fyre.ibm.com ibm-common-service-operator]# ./isolate.sh --original-cs-ns ibm-common-services --control-ns cs-control -v 1
/usr/bin/oc
/usr/local/bin/yq
[INFO] Checking for common service operator and odlm pods
[DEBUG] cs op scaled output: ibm-common-service-operator            1/1     1            1           17m
[INFO] Common Service Operator already scaled, skipping.
[DEBUG] odlm scaled output: operand-deployment-lifecycle-manager   1/1     1            1           16m
[INFO] ODLM already scaled, skipping.
[INFO] Verifying cert manager is deployed
[DEBUG] cert manager csv count output: 1
[✔] Cert manager deployment verified.
No resources found
No resources found in openshift-operators namespace.
# Pausing Common Services in namespace ibm-common-services
-----------------------------------------------------------------------
deployment.apps/ibm-common-service-operator scaled
deployment.apps/operand-deployment-lifecycle-manager scaled
operandregistry.operator.ibm.com "common-service" deleted
operandconfig.operator.ibm.com "common-service" deleted
[✔] Common Services successfully isolated in namespace ibm-common-services
[INFO] Deleting podpresets in namespace ibm-common-services...
podpreset.operator.ibm.com "ibm-common-service-webhook" deleted

[INFO] Deleting existing Deployment ibm-common-service-webhook in namespace ibm-common-services...
deployment.apps "ibm-common-service-webhook" deleted
[INFO] Deleting MutatingWebhookConfiguration...
mutatingwebhookconfiguration.admissionregistration.k8s.io "ibm-common-service-webhook-configuration" deleted

[INFO] Deleting ValidatingWebhookConfiguration...
validatingwebhookconfiguration.admissionregistration.k8s.io "ibm-cs-ns-mapping-webhook-configuration" deleted
#  Creating empty common-service-maps configmap 
configmap/common-service-maps created
[✔] Empty common-service-maps configmap created in kube-public namespace
configmap/common-service-maps configured
[INFO] The provided map-to-common-service-namespace: ibm-common-services, does not exist in common-service-maps
[INFO] Adding new map-to-common-service-namespace
[DEBUG] map ibm-common-services namespace ibm-common-services tmp "ibm-common-services"
[DEBUG] ns ibm-common-services mapto: ibm-common-services
[DEBUG] ns_delimited: "ibm-common-services"
configmap/common-service-maps configured
#  Verify all namespaces exist 
-----------------------------------------------------------------------
[INFO] Creating namespace ibm-common-services
Error from server (AlreadyExists): namespaces "ibm-common-services" already exists
[INFO] ibm-common-services already exists, skipping...
[INFO] Creating namespace cs-control
namespace/cs-control created
[✔] All namespaces in common-service-maps exist
#  Removing ODLM managed Namespace Scope CRs 
-----------------------------------------------------------------------
[INFO] deleting namespace scope nss-managedby-odlm in namespace ibm-common-services
namespacescope.operator.ibm.com "nss-managedby-odlm" deleted
[INFO] deleting namespace scope odlm-scope-managedby-odlm in namespace ibm-common-services
namespacescope.operator.ibm.com "odlm-scope-managedby-odlm" deleted
[INFO] deleting namespace scope nss-odlm-scope in namespace ibm-common-services
namespacescope.operator.ibm.com "nss-odlm-scope" deleted
[INFO] deleting namespace scope common-service in namespace ibm-common-services
namespacescope.operator.ibm.com "common-service" deleted
[✔] Namespace Scope CRs cleaned up
# Uninstalling Singleton Operators
-----------------------------------------------------------------------
certmanager.operator.ibm.com "default" deleted
[DEBUG] Cert Manager marked for migration.
subscription.operators.coreos.com "ibm-cert-manager-operator" deleted
clusterserviceversion.operators.coreos.com "ibm-cert-manager-operator.v3.25.3" deleted
# Copying over Licensing Configmaps
-----------------------------------------------------------------------
^C
[root@api.critter.cp.fyre.ibm.com ibm-common-service-operator]# oc get pods -A | grep cert
openshift-marketplace                              certified-operators-z9rzn                                         1/1     Running                      0             8h
openshift-marketplace                              ibm-cert-manager-catalog-glbd4                                    1/1     Running                      0             17h
openshift-marketplace                              ibm-cert-manager-operator-catalog-nphcj                           1/1     Running                      0             17h
[root@api.critter.cp.fyre.ibm.com ibm-common-service-operator]# oc get deploy | grep '0/0'
ibm-common-service-operator            0/0     0            0           17m
operand-deployment-lifecycle-manager   0/0     0            0           17m
[root@api.critter.cp.fyre.ibm.com ibm-common-service-operator]# git checkout scripts
Switched to branch 'scripts'
Your branch is behind 'origin/scripts' by 1 commit, and can be fast-forwarded.
  (use "git pull" to update your local branch)
[root@api.critter.cp.fyre.ibm.com ibm-common-service-operator]# git pull
Updating f59fa7e..992e16e
Fast-forward
 isolate.sh | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)
[root@api.critter.cp.fyre.ibm.com ibm-common-service-operator]# ./isolate.sh --original-cs-ns ibm-common-services --control-ns cs-control -v 1
/usr/bin/oc
/usr/local/bin/yq
[root@api.critter.cp.fyre.ibm.com ibm-common-service-operator]# git checkout 10076
Switched to branch '10076'
Your branch is up to date with 'origin/10076'.
[root@api.critter.cp.fyre.ibm.com ibm-common-service-operator]# ./isolate.sh --original-cs-ns ibm-common-services --control-ns cs-control -v 1
/usr/bin/oc
/usr/local/bin/yq
[INFO] Checking for common service operator and odlm pods
[DEBUG] cs op scaled output: false
deployment.apps/ibm-common-service-operator scaled
[INFO] Common Service Operator scaled back to 1
[DEBUG] odlm scaled output: false
deployment.apps/operand-deployment-lifecycle-manager scaled
[INFO] ODLM scaled back to 1
[INFO] Waiting for IBM Common Services CR is Succeeded
-----------------------------------------------------------------------
[✔] Ready use
Error from server (NotFound): configmaps "namespace-scope" not found
#  Wait for Cert Manager pods to come ready 
-----------------------------------------------------------------------
[INFO] Checking for cert manager deployed in scope.
[INFO] Cert manager requested in scope, moving on...
No resources found in null namespace.
[INFO] Waiting for deployment cert-manager-operator to be running ...
[INFO] RETRYING: Waiting for deployment cert-manager-operator to be running ... (20 left)
[INFO] RETRYING: Waiting for deployment cert-manager-operator to be running ... (19 left)
[INFO] RETRYING: Waiting for deployment cert-manager-operator to be running ... (18 left)
[INFO] RETRYING: Waiting for deployment cert-manager-operator to be running ... (17 left)
[✔] Deployment cert-manager-operator is running.
[INFO] Waiting for pod cert-manager-webhook to be running ...
[INFO] RETRYING: Waiting for pod cert-manager-webhook to be running ... (20 left)
[INFO] RETRYING: Waiting for pod cert-manager-webhook to be running ... (19 left)
[INFO] RETRYING: Waiting for pod cert-manager-webhook to be running ... (18 left)
[INFO] RETRYING: Waiting for pod cert-manager-webhook to be running ... (17 left)
[INFO] RETRYING: Waiting for pod cert-manager-webhook to be running ... (16 left)
[INFO] RETRYING: Waiting for pod cert-manager-webhook to be running ... (15 left)
[INFO] RETRYING: Waiting for pod cert-manager-webhook to be running ... (14 left)
[INFO] RETRYING: Waiting for pod cert-manager-webhook to be running ... (13 left)
[✔] Pod cert-manager-webhook is running.
[✔] Cert Manager ready. Cert Manager operands deployed in cs-control
[INFO] Verifying cert manager is deployed
[DEBUG] cert manager csv count output: 1
[✔] Cert manager deployment verified.
No resources found
No resources found in openshift-operators namespace.
# Pausing Common Services in namespace ibm-common-services
-----------------------------------------------------------------------
deployment.apps/ibm-common-service-operator scaled
deployment.apps/operand-deployment-lifecycle-manager scaled
operandregistry.operator.ibm.com "common-service" deleted
operandconfig.operator.ibm.com "common-service" deleted
[✔] Common Services successfully isolated in namespace ibm-common-services
No resources found in ibm-common-services namespace.
[INFO] Deleting existing Deployment ibm-common-service-webhook in namespace ibm-common-services...
[INFO] Deleting MutatingWebhookConfiguration...
mutatingwebhookconfiguration.admissionregistration.k8s.io "ibm-common-service-webhook-configuration" deleted
mutatingwebhookconfiguration.admissionregistration.k8s.io "ibm-operandrequest-webhook-configuration" deleted

[INFO] Deleting ValidatingWebhookConfiguration...
validatingwebhookconfiguration.admissionregistration.k8s.io "ibm-cs-ns-mapping-webhook-configuration" deleted
#  Creating empty common-service-maps configmap 
[INFO] The common-service-maps already exists, skipping
configmap/common-service-maps unchanged
[DEBUG] map ibm-common-services namespace ibm-common-services tmp "ibm-common-services"
[DEBUG] ns ibm-common-services mapto: ibm-common-services
[DEBUG] ns_delimited: "ibm-common-services"
configmap/common-service-maps unchanged
#  Verify all namespaces exist 
-----------------------------------------------------------------------
[INFO] Creating namespace ibm-common-services
Error from server (AlreadyExists): namespaces "ibm-common-services" already exists
[INFO] ibm-common-services already exists, skipping...
[INFO] Creating namespace cs-control
Error from server (AlreadyExists): namespaces "cs-control" already exists
[INFO] cs-control already exists, skipping...
[✔] All namespaces in common-service-maps exist
#  Removing ODLM managed Namespace Scope CRs 
-----------------------------------------------------------------------
[INFO] deleting namespace scope nss-managedby-odlm in namespace ibm-common-services
[INFO] deleting namespace scope odlm-scope-managedby-odlm in namespace ibm-common-services
[INFO] deleting namespace scope nss-odlm-scope in namespace ibm-common-services
namespacescope.operator.ibm.com "nss-odlm-scope" deleted
[INFO] deleting namespace scope common-service in namespace ibm-common-services
namespacescope.operator.ibm.com "common-service" deleted
[✔] Namespace Scope CRs cleaned up
# Uninstalling Singleton Operators
-----------------------------------------------------------------------
# Copying over Licensing Configmaps
-----------------------------------------------------------------------
[INFO] No licensing configmaps to migrate
[INFO] Licensing marked for backup
configmap/ibmlicensing-instance-bak created
[INFO] Deleting existing Deployment secretshare in namespace ibm-common-services...
deployment.apps "secretshare" deleted
[✔] Singletons successfully uninstalled
[DEBUG] patch string {"name": "INSTALL_SCOPE", "value": "namespaced"}, {"name": "ISOLATED_MODE", "value": "true"}, {"name": "OPERATORCHECKER_MODE", "value": "true"}
subscription.operators.coreos.com/operand-deployment-lifecycle-manager-app patched
[INFO] Waiting for OLM to update Deployment operand-deployment-lifecycle-manager 
[✔] Deployment operand-deployment-lifecycle-manager is updated
ibmlicensing.operator.ibm.com/instance configured
# Scaling up ibm-common-service-operator deployment in ibm-common-services namespace
-----------------------------------------------------------------------
deployment.apps/ibm-common-service-operator scaled
deployment.apps/operand-deployment-lifecycle-manager scaled
[INFO] Waiting for IBM Common Services CR is Succeeded
-----------------------------------------------------------------------
[✔] Ready use
[✔] Common Service Operator restarted.
#  Wait for Cert Manager pods to come ready 
-----------------------------------------------------------------------
[INFO] Checking for cert manager deployed in scope.
[INFO] Cert manager requested in scope, moving on...
[INFO] Waiting for deployment cert-manager-operator to be running ...
[✔] Deployment cert-manager-operator is running.
[INFO] Waiting for pod cert-manager-webhook to be running ...
[✔] Pod cert-manager-webhook is running.
[✔] Cert Manager ready. Cert Manager operands deployed in cs-control
[INFO] Waiting for configmap namespace-scope in namespace ibm-common-services to be created ...
[✔] Namespace-scope configmap created in ibm-common-services.
[DEBUG] expected ns list: ibm-common-services
[DEBUG] actual ns list: ibm-common-services
[✔] Namespaces in namespace-scope configmap match expected output.
[✔] Isolation complete
[root@api.critter.cp.fyre.ibm.com ibm-common-service-operator]# oc get pods -A | grep cert
cs-control                                         cert-manager-cainjector-5c9db88f64-v7v6r                          1/1     Running     0             7m43s
cs-control                                         cert-manager-controller-64bcf4cbdc-rfj4v                          1/1     Running     0             7m44s
cs-control                                         cert-manager-webhook-8b48ffc5b-r2m68                              1/1     Running     0             7m43s
cs-control                                         ibm-cert-manager-operator-57b44fd654-9stnq                        1/1     Running     0             8m58s
openshift-marketplace                              certified-operators-z9rzn                                         1/1     Running     0             8h
openshift-marketplace                              ibm-cert-manager-catalog-glbd4                                    1/1     Running     0             17h
openshift-marketplace                              ibm-cert-manager-operator-catalog-nphcj                           1/1     Running     0             17h
[root@api.critter.cp.fyre.ibm.com ibm-common-service-operator]# 
```

bluzarraga commented:
> @bluzarraga hmmm I doubt that we do actually need `echo "none"`? My assumption is that we will always have Kind `CSV` existing in the cluster, there is no non-zero status returned from first command
> 
> ```
> ➜  csv_count=$(oc get csv -A | grep "cert-manager"| wc -l | tr -d " ")     
> ➜  echo $?
> 0
> ➜  echo $csv_count
> 2
> ➜  csv_count=$(oc get csv -A | grep "cert-manager-xxxx"| wc -l | tr -d " ")
> ➜  echo $?                                                                 
> 0
> ➜  echo $csv_count                                                         
> 0
> ```

I agree, we should not be outputting `none` because I was seeing the csv_count var equal to `0\nnone` which is not what we are looking for. I have changed it to echo "" instead. I still think we should echo something because the quick error we are seeing means that that command is failing and immediately exiting instead of getting a 0 value and then failing the subsequent check. That's why we aren't seeing the error message. I added the debug statement right below `csv_count=$(${OC} get csv -A | grep "cert-manager"| wc -l | tr -d " ")` while running on one of the environments from the issue and it did not output anything before failing.

bitscuit commented:
Tested changes on my cluster, and everything seemed good

Daniel-Fan commented on isolate.sh:
```diff
@@ -736,8 +738,10 @@ function wait_for_certmanager() {
 }
 
 function check_certmanager_count(){
```
```suggestion
function check_certmanager_count(){
    info "Verifying cert manager is deployed"
    local webhook_deployments=$(${OC} get deploy -A --no-headers --ignore-not-found | grep "cert-manager-webhook" | wc -l | tr -d " ")
    debug1 "cert manager webhook deployment count output: $webhook_deployments"

    if [[ "$webhook_deployments" == "0" ]]; then
        error "Missing a cert-manager"
    elif [[ "$webhook_deployments" == "1" ]]; then
        success "Cert manager deployment verified."
    else
        error "Multiple cert-managers found. Only one should be installed per cluster"
    fi
}
```

Daniel-Fan reviewed (approved):
/lgtm
